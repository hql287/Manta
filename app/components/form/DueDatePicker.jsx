// Libraries
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';

// React Dates
import { SingleDatePicker } from 'react-dates';

// Styles
import _withFadeInAnimation from '../shared/hoc/_withFadeInAnimation';
import styled from 'styled-components';
const Container = styled.div`
  display: flex;
  margin-bottom: 20px;
`;

// Component
export class DueDatePicker extends PureComponent {
  constructor(props) {
    super(props);
    this.state = { focused: false };
    this.onFocusChange = this.onFocusChange.bind(this);
    this.onDateChange = this.onDateChange.bind(this);
    this.clearDate = this.clearDate.bind(this);
  }

  onFocusChange() {
    this.setState({ focused: !this.state.focused });
  }

  onDateChange(date) {
    const selectedDate = date === null ? null : moment(date).toObject();
    this.props.updateCustomDate(selectedDate);
  }

  clearDate() {
    this.onDateChange(null);
  }

  render() {
    const { t, selectedDate } = this.props;
    const dueDate = selectedDate === null ? null : moment(selectedDate);
    return (
      <Container>
        <SingleDatePicker
          id="invoice-duedate"
          placeholder={t('form:fields:dueDate:placeHolder')}
          firstDayOfWeek={1}
          withFullScreenPortal
          displayFormat="DD/MM/YYYY"
          hideKeyboardShortcutsPanel
          date={dueDate}
          focused={this.state.focused}
          onFocusChange={this.onFocusChange}
          onDateChange={newDate => this.onDateChange(newDate)}
        />
        {selectedDate !== null && (
          <a className="clearDateBtn active" href="#" onClick={this.clearDate}>
            <i className="ion-close-circled" />
          </a>
        )}
      </Container>
    );
  }
}

DueDatePicker.propTypes = {
  selectedDate: PropTypes.object,
  t: PropTypes.func.isRequired,
  updateCustomDate: PropTypes.func.isRequired,
};

DueDatePicker.defaultProps = {
  selectedDate: null,
};

// Export
export default _withFadeInAnimation(DueDatePicker);
