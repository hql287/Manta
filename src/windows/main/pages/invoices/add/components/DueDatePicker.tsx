// Libraries
import React, { useState } from 'react'
import DatePicker from 'react-datepicker'
import 'react-datepicker/dist/react-datepicker.css'
import styled from 'styled-components'
import { format, parseISO } from 'date-fns'

// Styles
import _withFadeInAnimation from '@uiSharedComponents/hoc/_withFadeInAnimation'

const Container = styled.div`
  display: flex;
  margin-bottom: 20px;
`

const ClearDateButton = styled.a`
  margin-left: 10px;
  color: red;
  cursor: pointer;

  &:hover {
    color: darkred;
  }

  i {
    font-size: 20px;
  }
`

// Component
interface DueDatePickerProps {
  selectedDate: string | null
  t: (key: string) => string
  updateCustomDate: (date: string | null) => void
}

const DueDatePicker: React.FC<DueDatePickerProps> = ({
  selectedDate,
  t,
  updateCustomDate,
}) => {
  const [focused, setFocused] = useState(false)

  const handleDateChange = (date: Date | null) => {
    const formattedDate = date ? format(date, 'yyyy-MM-dd') : null
    updateCustomDate(formattedDate)
  }

  const clearDate = () => {
    handleDateChange(null)
  }

  const parsedDate = selectedDate ? parseISO(selectedDate) : null

  return (
    <Container>
      <DatePicker
        selected={parsedDate}
        onChange={handleDateChange}
        placeholderText={t('form:fields:dueDate:placeHolder')}
        dateFormat="dd/MM/yyyy"
        onFocus={() => setFocused(true)}
        onBlur={() => setFocused(false)}
      />
      {selectedDate && (
        <ClearDateButton onClick={clearDate}>
          <i className="ion-close-circled" />
        </ClearDateButton>
      )}
    </Container>
  )
}

export default _withFadeInAnimation(DueDatePicker)
