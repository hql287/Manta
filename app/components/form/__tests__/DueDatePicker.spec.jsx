// Libs
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { SingleDatePicker } from 'react-dates';

// Component
import { DueDatePicker } from '../DueDatePicker.jsx';

// Mocks
const t = jest.fn();
const updateCustomDate = jest.fn();
const selectedDate = {
  milliseconds: 0,
  seconds: 0,
  minutes: 0,
  hours: 12,
  date: 20,
  months: 10,
  years: 2017,
};

describe('Renders correctly to the DOM', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(
      <DueDatePicker t={t} selectedDate={selectedDate} updateCustomDate={updateCustomDate} />
    );
  });

  // TODO
  it('will not rerender if threre is no change in Props', () => {});

  it('receives correct props', () => {
    expect(wrapper.prop('selectedDate')).toEqual(selectedDate);
    expect(wrapper.prop('selectedDate')).not.toEqual({});
    expect(wrapper.prop('updateCustomDate')).toEqual(updateCustomDate);
  });

  it('renders necessary element', () => {
    expect(wrapper.find(SingleDatePicker)).toHaveLength(1);
    expect(wrapper.find(SingleDatePicker)).not.toHaveLength(2);
  });

  it('call clearDate when click clearDate Button', () => {
    // Setup
    const spy = jest.spyOn(DueDatePicker.prototype, 'clearDate');
    const wrap = mount(
      <DueDatePicker t={t} selectedDate={selectedDate} updateCustomDate={updateCustomDate} />
    );
    const clearDateBtn = wrap.find('.clearDateBtn');
    // Execute
    clearDateBtn.simulate('click');
    // Assertion
    expect(spy).toHaveBeenCalled();
  });

  it('clearDate clears selectedDate', () => {
    // Setup
    const instance = wrapper.instance();
    const spy = jest.spyOn(instance, 'clearDate');
    // Execute
    instance.clearDate();
    // Assertion
    expect(spy).toHaveBeenCalled();
    expect(updateCustomDate).toHaveBeenCalled();
    expect(updateCustomDate).toHaveBeenCalledWith(null);
  });

  it('matches snapshot', () => {
    const tree = renderer
      .create(
        <DueDatePicker t={t} selectedDate={selectedDate} updateCustomDate={updateCustomDate} />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
