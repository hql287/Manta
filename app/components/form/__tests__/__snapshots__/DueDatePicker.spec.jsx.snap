// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly to the DOM matches snapshot 1`] = `
<div
  className="DueDatePicker__Container-jrhXSv dVnPcX"
>
  <div
    className="SingleDatePicker"
  >
    <div>
      <div
        className="SingleDatePickerInput"
      >
        <div
          className="DateInput DateInput--open-down"
        >
          <input
            aria-describedby="DateInput__screen-reader-message-invoice-duedate"
            aria-label="Date"
            autoComplete="off"
            className="DateInput__input needsclick"
            disabled={false}
            id="invoice-duedate"
            name="invoice-duedate"
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Date"
            readOnly={false}
            required={false}
            type="text"
            value="2017-11-20"
          />
          <p
            className="screen-reader-only"
            id="DateInput__screen-reader-message-invoice-duedate"
          >
            Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
          </p>
          <div
            className="DateInput__display-text DateInput__display-text--has-input"
          >
            20/11/2017
          </div>
        </div>
      </div>
    </div>
  </div>
  <a
    className="clearDateBtn active"
    href="#"
    onClick={[Function]}
  >
    <i
      className="ion-close-circled"
    />
  </a>
</div>
`;
